### PR TITLE
Support show examples for azure management sdk in sdk reference book #7095

### PR DESCRIPTION
**Pull request by Flanker32**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Support show examples for azure management sdk in sdk reference book

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------
…

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/AzureJavaSdkArtifactExampleIndexEntity.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @JsonProperty("language")
+    private String language;
+    @JsonProperty("tag")
+    private String tag;
+    @JsonProperty("package")
+    private String packageName;
+    @JsonProperty("version")
+    private String version;
+    @JsonProperty("date_epoch")
+    private String date_epoch;
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/AzureJavaSdkArtifactExampleIndexEntity.java:31]: dateEpoch

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/module/GradleProjectModule.java [hunk 2]
```diff
@@ -47,6 +48,14 @@ public ExternalDependency getGradleDependency(final String groupId, final String
                 .filter(dependency -> !(dependency instanceof UnresolvedExternalDependency)).findFirst().orElse(null);
     }
 
+    @Override
+    public ComparableVersion getDependencyVersion(@Nullable AzureSdkArtifactEntity entity) {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/module/GradleProjectModule.java:52]: don't we have a shared `GradleProjectModule`?
this class should not depends on `AzureSdkArtifactEntity`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/referencebook/AzureSdkArtifactGroupPanel.java [hunk 3]
```diff
@@ -94,6 +108,11 @@ private void onPackageOrVersionSelected(AzureSdkArtifactEntity pkg, String versi
         this.pnlAddDependencies.setPkg(pkg);
         this.pnlAddDependencies.setVersion(version);
         this.pnlAddDependencies.onSelectModule();
+
+        final AzureJavaSdkArtifactExampleIndexEntity sdkExampleIndex = AzureSdkExampleService.getSdkExampleIndex(pkg);
+        final List<AzureJavaSdkArtifactExampleEntity> examples = Optional.ofNullable(sdkExampleIndex)
+                .map(AzureJavaSdkArtifactExampleIndexEntity::getExamples).orElse(null);
+        this.pnlDependency.setExampleIndex(sdkExampleIndex);
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/referencebook/AzureSdkArtifactGroupPanel.java:115]: it's strange to set an **index** here.
it's strange to set a example to a dependency panel

Comment (wangmingliang-ms) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/referencebook/AzureSdkArtifactGroupPanel.java:115]: why not just get examples based on the selected artifact and then set them to the examples panel?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    private static final String JAVA_LIBRARY_EXAMPLE_INDEX_CSV = "/java-library-example-index.csv";
+    private static final Map<String, String> TEMPLATE_CACHE = new HashMap<>();
+
+    public static String loadSdkTemplate(@Nonnull final AzureJavaSdkArtifactExampleEntity entity) {
+        return TEMPLATE_CACHE.computeIfAbsent(entity.getRawUrl(), ignore -> {
+            try {
+                return IOUtils.toString(new URL(entity.getRawUrl()), Charset.defaultCharset());
+            } catch (final IOException e) {
+                log.warn("failed to load Azure SDK example", e);
+                return null;
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:49]: Nullable

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+@Slf4j
+public class AzureSdkExampleService {
+    private static final ObjectMapper CSV_MAPPER = new CsvMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    private static final String JAVA_LIBRARY_EXAMPLE_LIST_CSV = "/java-library-example-list.csv";
+    private static final String JAVA_LIBRARY_EXAMPLE_INDEX_CSV = "/java-library-example-index.csv";
+    private static final Map<String, String> TEMPLATE_CACHE = new HashMap<>();
+
+    public static String loadSdkTemplate(@Nonnull final AzureJavaSdkArtifactExampleEntity entity) {
+        return TEMPLATE_CACHE.computeIfAbsent(entity.getRawUrl(), ignore -> {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:44]: why not use `Cacheable`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    public static AzureJavaSdkArtifactExampleIndexEntity getSdkExampleIndex(@Nonnull AzureSdkArtifactEntity entity) {
+        return loadAzureSDKExample().stream()
+                .filter(example -> StringUtils.equalsIgnoreCase(example.getGroupId(), entity.getGroupId())
+                        && StringUtils.equalsIgnoreCase(example.getPackageName(), entity.getArtifactId()))
+                .findFirst().orElse(null);
+    }
+
+    @Preload
+    @Cacheable(value = "sdk/examples")
+    public static List<AzureJavaSdkArtifactExampleIndexEntity> loadAzureSDKExample() {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:64]: `loadAllExamples`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                return IOUtils.toString(new URL(entity.getRawUrl()), Charset.defaultCharset());
+            } catch (final IOException e) {
+                log.warn("failed to load Azure SDK example", e);
+                return null;
+            }
+        });
+    }
+
+    @Nullable
+    public static AzureJavaSdkArtifactExampleIndexEntity getSdkExampleIndex(@Nonnull AzureSdkArtifactEntity entity) {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:55]: getArtifactExamples

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        final List<AzureJavaSdkArtifactExampleIndexEntity> indexEntities = loadAzureSDKExampleIndex().stream()
+                .filter(entity -> StringUtils.equalsIgnoreCase(entity.getLanguage(), "java"))
+                .collect(Collectors.toList());
+        indexEntities.forEach(entity -> entity.setExamples(exampleMap.get(entity.getId())));
+        return indexEntities;
+    }
+
+    @Cacheable(value = "java-library-example-index")
+    @AzureOperation(name = "sdk.load_library_example_index", type = AzureOperation.Type.TASK, target = AzureOperation.Target.PLATFORM)
+    public static List<AzureJavaSdkArtifactExampleIndexEntity> loadAzureSDKExampleIndex() {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:76]: does this method have to be `public`?
why does this method have same return types but different names, how can user determine which method to use?

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                .collect(Collectors.groupingBy(e -> e.getReleaseId(), Collectors.mapping(e -> e, Collectors.toList())));
+        final List<AzureJavaSdkArtifactExampleIndexEntity> indexEntities = loadAzureSDKExampleIndex().stream()
+                .filter(entity -> StringUtils.equalsIgnoreCase(entity.getLanguage(), "java"))
+                .collect(Collectors.toList());
+        indexEntities.forEach(entity -> entity.setExamples(exampleMap.get(entity.getId())));
+        return indexEntities;
+    }
+
+    @Cacheable(value = "java-library-example-index")
+    @AzureOperation(name = "sdk.load_library_example_index", type = AzureOperation.Type.TASK, target = AzureOperation.Target.PLATFORM)
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:75]: `REQUEST`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            final MappingIterator<AzureJavaSdkArtifactExampleIndexEntity> data = reader.readValues(stream);
+            return data.readAll().stream().collect(Collectors.toList());
+        } catch (final IOException e) {
+            log.warn("failed to load Azure SDK example index", e);
+        }
+        return Collections.emptyList();
+    }
+
+    @Cacheable(value = "java-library-example")
+    @AzureOperation(name = "sdk.load_library_example", type = AzureOperation.Type.TASK, target = AzureOperation.Target.PLATFORM)
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:88]: `REQUEST`

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    public static List<AzureJavaSdkArtifactExampleEntity> loadAzureSDKExampleEntities() {
```
Comment (wangmingliang-ms) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:89]: public?

**Review by wangmingliang-ms — APPROVED** [review]
LGTM

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/module/GradleProjectModule.java (hunk 2 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/model/module/GradleProjectModule.java:52]: Currently we don't have a shared project module, we may add one in later commits

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/referencebook/AzureSdkArtifactGroupPanel.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/referencebook/AzureSdkArtifactGroupPanel.java:115]: dependencyPanel should be examplePanel
change to setArtifact, and example panel will get artifact from service

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java (hunk 9 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:75]: This is not a web request, but read from local meta data

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java (hunk 10 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:88]: This is not a web request, but from local meta data

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java (hunk 11 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:89]: Changed to private

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java (hunk 8 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/AzureSdkExampleService.java:76]: rename to `loadAzureSDKExamplesEntities`

Here we just load the `ExamplesEntity`, in `loadAllExamples`, we will also load the detail `ExampleEntity` and set it to `ExamplesEntity`

**Pull request merged**